clang/include/clang/Basic/AttrDocs.td

```diff
@@ -8119,6 +8119,35 @@ but do not pass them to the underlying coroutine or pass them by value.
 }];
 }
 
+def CoroAwaitElidableDoc : Documentation {
+  let Category = DocCatDecl;
+  let Content = [{
+The ``[[clang::coro_await_elidable]]`` is a class attribute which can be applied
+to a coroutine return type.
+
+When a coroutine function that returns such a type calls another coroutine function,
+the compiler performs heap allocation elision when the call to the coroutine function
+is immediately co_awaited as a prvalue.
```

Comment:
What I am concerning is, actually we don't care if it is prvalue or not at all. What we really want is to find the CallInst. The prvalue catagory doesn't be involved in the design. It is just, the idea of `prvalue` can help us to implement this. So I think the concept of prvalue is better to be an implementation details but not the interface.

And on the other hand, in my mind, this should be meant to be related with parsers. So I am fine with that.

Reply:
> What I am concerning is, actually we don't care if it is prvalue or not at all.

I am actually using `isPRValue()` in Sema for this case. The right-hand side of `co_await` being a prvalue is a crucial requirement of this design. If the right-hand side is not a prvalue, we cannot do the elide safely.

Reply:
This is about the implementation design. What I am saying is about the mental interface to users.

>  If the right-hand side is not a prvalue, we cannot do the elide safely.

Then the question is, if the call is immediately co_awaited, in what cases it won't be a prvalue. And if it is, why it is not safe on the high level?

Reply:
It still did not sound right to me. I was trying to understand the term "immediate" in the context of syntactic analysis. It's defined to be a different meaning in the standard. 

https://eel.is/c++draft/intro.execution#3

Of course, it's not the first time we use an overloaded term and it won't be the last.

Reply:
> Then the question is, if the call is immediately co_awaited, in what cases it won't be a prvalue.

There probably aren't any. But I was still not confident of what "immediately" means in general. 

The term "immediate subexpression" is defined to be a different meaning in the standard: https://eel.is/c++draft/intro.execution#3

But I don't have a strong opinion of this. Probably not ambiguous in this context. This isn't the first time we use an overloaded term and won't be the last either, but I still like interfacing with developer with the term `prvalue` better. In my experience it's better understood.

Reply:
Yeah, the term "immediate subexpression" is not clear in the standard wording. But we're using a different term "immediately co_awaited". So we're not covering something existed already. The reason why I don't like `prvalue` here is that it didn't interpret the things very clearly. It may make people think they understand it but in fact it is not.

In another word, the semantics or intention of the attribute is that, the coroutine instance generated by the targeted call (no matter how we describe it) before we resume the current coroutine. The interesting point here is that, the coroutine instance is not necessarily connected to the return object of that targeted call. They are two different things. But just in the general implementation, the return object of such targeted call may have a pointer pointed to the coroutine instance. But there are indeed differences.